vpr/src/place/place.cpp

```diff
@@ -1085,48 +1090,61 @@ static void update_rlim(float* rlim, float success_rat, const DeviceGrid& grid)
 }
 
 /* Update the temperature according to the annealing schedule selected. */
-static void update_t(float* t, float rlim, float success_rat, t_annealing_sched annealing_sched) {
-    /*  float fac; */
-
+static bool update_state(t_annealing_state* state,
+                         float success_rat,
+                         const t_placer_costs& costs,
+                         const t_placer_opts& placer_opts,
+                         const t_annealing_sched& annealing_sched) {
+    /* Return `false` when the exit criterion is met. */
     if (annealing_sched.type == USER_SCHED) {
-        *t = annealing_sched.alpha_t * (*t);
-    } else { /* AUTO_SCHED */
-        if (success_rat > 0.96) {
-            *t = (*t) * 0.5;
-        } else if (success_rat > 0.8) {
-            *t = (*t) * 0.9;
-        } else if (success_rat > 0.15 || rlim > 1.) {
-            *t = (*t) * 0.95;
-        } else {
-            *t = (*t) * 0.8;
-        }
+        state->t *= annealing_sched.alpha_t;
+        return state->t >= annealing_sched.exit_t;
     }
-}
 
-static int exit_crit(float t, float cost, t_annealing_sched annealing_sched) {
-    /* Return 1 when the exit criterion is met.                        */
+    auto& device_ctx = g_vpr_ctx.device();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
 
-    if (annealing_sched.type == USER_SCHED) {
-        if (t < annealing_sched.exit_t) {
-            return (1);
+    /* Automatic annealing schedule */
+    float t_exit = 0.005 * costs.cost / cluster_ctx.clb_nlist.nets().size();
+
+    if (annealing_sched.type == DUSTY_SCHED) {
```

Comment:
I'm wondering if the annealing schedule should be behind an interface that consumes the state?  Or would that be too slow?

Reply:
That would be the right thing to do in C++-style code. Functionally, it's probably the same. I decided to avoid changing the style in `place.cpp` to avoid a full rewrite.